drivers/regulator/regulator_npm6001.c

```diff
@@ -0,0 +1,693 @@
+	const struct linear_range *range, uint8_t vout_reg, uint8_t conf_reg)
+{
+	const struct regulator_npm6001_config *config = dev->config;
+	const struct regulator_npm6001_pconfig *pconfig = config->p->config;
+	uint8_t conf, buf[3];
+	uint16_t idx;
+	int ret;
+
+	ret = linear_range_get_win_index(range, min_uv, max_uv, &idx);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* force PWM mode when updating voltage */
+	ret = i2c_reg_read_byte_dt(&pconfig->i2c, conf_reg, &conf);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if ((conf & NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM) == 0U) {
+		ret = i2c_reg_write_byte_dt(
+			&pconfig->i2c, conf_reg,
+			conf | NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	/* write voltage in both ULP/PWM registers */
+	buf[0] = vout_reg;
+	buf[1] = (uint8_t)idx;
+	buf[2] = (uint8_t)idx;
+
+	ret = i2c_write_dt(&pconfig->i2c, buf, sizeof(buf));
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = i2c_reg_write_byte_dt(&pconfig->i2c, NPM6001_TASKS_UPDATE_VOUTPWM,
+				    1U);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* restore HYS mode if enabled before */
+	if ((conf & NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM) == 0U) {
+		ret = i2c_reg_write_byte_dt(&pconfig->i2c, conf_reg, conf);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int regulator_npm6001_buck3_set_voltage(const struct device *dev,
+					       int32_t min_uv, int32_t max_uv)
+{
+	const struct regulator_npm6001_config *config = dev->config;
+	const struct regulator_npm6001_pconfig *pconfig = config->p->config;
+	uint16_t idx;
+	uint8_t conf;
+	int ret;
+
+	ret = linear_range_get_win_index(&buck3_range, min_uv, max_uv, &idx);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* force PWM mode when updating voltage */
+	ret = i2c_reg_read_byte_dt(&pconfig->i2c, NPM6001_BUCK3CONFPWMMODE,
+				   &conf);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if ((conf & NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM) == 0U) {
+		ret = i2c_reg_write_byte_dt(
+			&pconfig->i2c, NPM6001_BUCK3CONFPWMMODE,
+			conf | NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	ret = i2c_reg_write_byte_dt(&pconfig->i2c, NPM6001_BUCK3VOUT,
+				    (uint8_t)idx);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* restore HYS mode if enabled before */
+	if ((conf & NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM) == 0U) {
+		ret = i2c_reg_write_byte_dt(&pconfig->i2c,
+					    NPM6001_BUCK3CONFPWMMODE, conf);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int regulator_npm6001_ldo0_set_voltage(const struct device *dev,
+					      int32_t min_uv, int32_t max_uv)
+{
+	const struct regulator_npm6001_config *config = dev->config;
+	const struct regulator_npm6001_pconfig *pconfig = config->p->config;
+	uint8_t val;
+
+	if ((min_uv <= 1800000) && (max_uv >= 1800000)) {
+		val = NPM6001_LDO0VOUT_SET1V8;
+	} else if ((min_uv > 1800000) && (min_uv <= 2100000)
+		   && (max_uv >= 2100000)) {
+		val = NPM6001_LDO0VOUT_SET2V1;
```

Comment:
This `(min_uv > 1800000)` condition is redundant because the previous `if` catches all values filtered out by it. For all voltages, it is sufficient to only check if min <= v and max >= v. It could be actually done in a loop utilizing the `ldo0_voltages` array and another one with `NPM6001_LDO0VOUT_*` values. Those two could be then also used in `regulator_npm6001_ldo0_get_voltage()` and we'd have all LDO0 voltage values specified in one place (just a suggestion, I won't insist :wink:).

Reply:
good suggestion let me improve the code. I'm really frustrated with the design of LDO0 registers, why were'nt they written as a linear range...

Reply:
@anangl please check latest patch